main.go

```diff
@@ -0,0 +1,175 @@
+	mu   sync.Mutex
+}
+
+func main() {
+	purchaseEvents := make(chan eth.Event)
+	stateEvents := make(chan db.Org)
+	go terminateExpiringOrgs(stateEvents)
+	var lastTopUp big.Int
+	lastTopUp.SetUint64(s.last["LastTopUp"])
+	lastTopUp.Add(&lastTopUp, big.NewInt(1))
+	go eth.StartListening(purchaseEvents, &lastTopUp)
+	go eth.UpdateCurrentBlock(&currentBlock)
+
+	for {
+		// get purchase events on eth contract
+		e := <-purchaseEvents
+
+		go func() {
+			// insert org, update expiry if exists
+			if db.UpsertOrg(e) {
+				// if UpsertOrg returns true org existed
+				// and we can skip the following steps
+				// TODO: if status != initial, we also need provider (retry case)
+				increaseWatermark("LastTopUp", e.BlockNumber)
+				return
+			}
+
+			// reserve a server
+			provider, ip, err := cloud.ReserveServer(e.Org)
+			if err != nil {
+				l.Println("Couldn't reserve server", err)
+
+			}
+			//provider, ip := "hetzner", "127.0.0.1"
+
+			// update ip for org
+			if err := db.UpdateOrgServer(e.Org, ip, provider); err != nil {
+				l.Println("Failed updating ip for org", e, err)
+			}
+
+			// update dns records for radicle subdomain
+			cloud.UpdateDNS(e.Org, ip)
+
+			// create .key file for org
+			utils.GenerateIdentity(e.Org)
+
+			// deploy ansible for initial setup
+			if err := cloud.RunAnsible(e.Org, ip, 10); err != nil {
+				// failed aftter 10 retries
+				db.SetStatus(e.Org, "setup-failed")
+			} else {
+				// flip status to runnning
+				db.SetStatus(e.Org, "running")
+				stateEvents <- db.Org{Org: e.Org, Expiry: e.Expiry, Provider: provider}
+				// set watermark for LastTopUp
+				// NOTE: race condition if multiple orders are on the
+				//		 same block. can lead to failure if one goes
+				//		 updating watermark and others fail
+				increaseWatermark("LastTopUp", e.BlockNumber)
+			}
+		}()
+	}
+}
+
+func setup() {
+	err := godotenv.Load()
+	if err != nil {
+		log.Fatal("Error loading .env file")
+	}
+
+	db.Setup()
+	cloud.Setup()
+
+	s = &procesingState{
+		last: make(map[string]uint64),
+	}
+	res, err := db.ListWatermarks()
+	if err != nil {
+		l.Panicln("Failed to get watermarks", err)
+	}
+	s.last = res
+}
+
+func increaseWatermark(watermark string, value uint64) error {
+	// first update in db, on success update global state
+	err := db.SetWatermark(watermark, value)
+	if err != nil {
+		return err
+	}
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if value > s.last[watermark] {
+		s.last[watermark] = value
+	}
+	return nil
+}
+
+func terminateExpiringOrgs(c chan db.Org) {
+	// update expiries with live events
+	var updatedExpiry map[string]uint64
+	// list all orgs with ascending expiring date
+	orgs, err := db.ListOrgs()
+	if err != nil {
+		l.Fatal("Can't list Orgs and Expiries")
+	}
+	for {
+		// terminate anything that has expired
+		for {
+			inext, ok := orgs.Peek()
+			if !ok {
+				// break out of loop if there's nothing
+				break
+			}
+			next := inext.(db.Org)
+			if next.Expiry <= currentBlock {
```

Comment:
Can't we query the DB for all orgs with an expiry that is `<=` currentBlock

Reply:
Yes, but not as efficient as current implementation. If we ever decide to use a cloud based db, we'll have to pay per reads/calls! Currently we only read once and then update global state via live events.

Reply:
Gotcha.